qemu/tests/smt_test.py

```diff
@@ -0,0 +1,120 @@
+import logging
+import re
+import time
+
+from virttest import env_process
+from virttest import error_context
+from virttest import utils_misc
+from virttest import utils_test
+
+from virttest.utils_test import BackgroundTest
+from provider.cpu_utils import check_cpu_flags
+
+
+@error_context.context_aware
+def run(test, params, env):
+    """
+    smt test:
+    1) Check if host has topoext flag, if not, cancel test
+    2) Boot guest and check cpu count and threads number in guest
+    3) Run stress inside guest, and check cpu usage in guest(only for linux)
+
+    :params test: QEMU test object.
+    :params params: Dictionary with the test parameters.
+    :params env: Dictionary with test environment.
+    """
+    def get_guest_threads():
+        """
+        Get guest threads number
+        """
+        if os_type == "linux":
+            cmd = params["get_threads_cmd"]
+            output = session.cmd_output_safe(cmd)
+            threads = int(re.findall(r":\s*(\d+)", output)[0])
+        else:
+            cmd = params["get_cores_cmd"]
+            output = session.cmd_output_safe(cmd)
+            cores = int(re.findall(r"=(\d+)", output)[0])
+            cmd = params["get_sockets_cmd"]
+            output = session.cmd_output_safe(cmd)
+            sockets = len(re.findall(r"SocketDesignation=", output))
+            threads = int(vm.cpuinfo.smp/sockets/cores)
+        return threads
+
+    def heavyload_install(install_path):
+        """
+        Install heavyload in windows guest
+        """
+        test_installed_cmd = 'dir "%s" | findstr /I heavyload' % install_path
+        if session.cmd_status(test_installed_cmd) != 0:
+            logging.warning("Could not find installed heavyload in guest, will"
+                            " install it via winutils.iso ")
+            winutil_drive = utils_misc.get_winutils_vol(session)
+            if not winutil_drive:
+                test.cancel("WIN_UTILS CDROM not found.")
+            install_cmd = params["install_cmd"] % winutil_drive
+            session.cmd(install_cmd)
+
+    def run_stress():
+        """
+        Run stress inside guest, return guest cpu usage
+        """
+        error_context.context("Run stress in guest and get cpu usage",
+                              logging.info)
+        if os_type == "linux":
+            stress_args = params["stress_args"]
+            stress_test = utils_test.VMStress(vm, "stress",
+                                              params, stress_args=stress_args)
+            try:
+                stress_test.load_stress_tool()
+                time.sleep(stress_duration / 2)
+                output = session.cmd_output_safe(params["get_cpu_usage_cmd"])
+                utils_misc.wait_for(lambda: (stress_test.app_running is False), 30)
+                stress_test.unload_stress()
+                cpu_usage = re.findall(r":\s*(\d+.?\d+)\s*us", output)
```

Comment:
Is this what you want? Or it is different on AMD host?
```
>>> output = subprocess.Popen('top -1 -n 1 | grep Cpu', shell=True, stdout=subprocess.PIPE).stdout.read().decode()
>>> print(output)
%Cpu0  :  0.0 us,  0.0 sy,  0.0 ni, 94.4 id,  0.0 wa,  5.6 hi,  0.0 si,  0.0 st
%Cpu1  :  5.9 us, 11.8 sy,  0.0 ni, 82.4 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
%Cpu2  :  5.9 us,  0.0 sy,  0.0 ni, 94.1 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
%Cpu3  :  5.3 us,  0.0 sy,  0.0 ni, 89.5 id,  0.0 wa,  5.3 hi,  0.0 si,  0.0 st
%Cpu4  :  0.0 us,  0.0 sy,  0.0 ni,100.0 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
%Cpu5  :  0.0 us,  0.0 sy,  0.0 ni,100.0 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
%Cpu6  :  0.0 us,  0.0 sy,  0.0 ni,100.0 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
%Cpu7  : 11.1 us,  0.0 sy,  0.0 ni, 88.9 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st

>>> re.findall(r":\s*(\d+.?\d+)\s*us", output)
[]
```

Reply:
Yes, it intends to get the value before 'us' for each line of output. And it's same whether it's amd or intel host.
In your cmd, output contains some ANSI escape code, that's why re.findall returns empty.
>>> output = subprocess.Popen('top -1 -n 1 | grep Cpu', shell=True, stdout=subprocess.PIPE).stdout.read().decode()
>>> output
'%Cpu0  :\x1b(B\x1b[m\x1b[39;49m\x1b[1m  0.0 \x1b(B\x1b[m\x1b[39;49mus,\x1b(B\x1b[m\x1b[39;49m\x1b[1m  0.0 \x1b(B\x1b[m\x1b[39;49msy,\x1b(B\x1b[m\x1b[39;49m\x1b[1m  0.0 \x1b(B\x1b[m\x1b[39;49mni,\x1b(B\x1b[m\x1b[39;49m\x1b[1m100.0 \x1b(B\x1b[m\x1b[39;49mid,\x1b(B\x1b[m\x1b[39;49m\x1b[1m  0.0 \x1b(B\x1b[m\x1b[39;49mwa,\x1b(B\x1b[m\x1b[39;49m\x1b[1m  0.0 \x1b(B\x1b[m\x1b[39;49mhi,\x1b(B\x1b[m\x1b[39;49m\x1b[1m  0.0 \x1b(B\x1b[m\x1b[39;49msi,\x1b(B\x1b[m\x1b[39;49m\x1b[1m  0.0 \x1b(B\x1b[m\x1b[39;49mst\x1b(B\x1b[m\x1b[39;49m\x1b(B\x1b[m\x1b[39;49m\x1b[K\n%Cpu1  :\x1b(B\x1b[m\x1b[39;49m\x1b[1m  0.0 \x1b(B\x1b[m\x1b[39;49mus,\x1b(B\x1b[m\x1b[39;49m\x1b[1m  0.0 \x1b(B\x1b[m\x1b[39;49msy,\x1b(B\x1b[m\x1b[39;49m\x1b[1m  0.0 \x1b(B\x1b[m\x1b[39;49mni,\x1b(B\x1b[m\x1b[39;49m\x1b[1m100.0 \x1b(B\x1b[m\x1b[39;49mid,\x1b(B\x1b[m\x1b[39;49m\x1b[1m  0.0 \x1b(B\x1b[m\x1b[39;49mwa,\x1b(B\x1b[m\x1b[39;49m\x1b[1m  0.0 \x1b(B\x1b[m\x1b[39;49mhi,\x1b(B\x1b[m\x1b[39;49m\x1b[1m  0.0 \x1b(B\x1b[m\x1b[39;49msi,\x1b(B\x1b[m\x1b[39;49m\x1b[1m  0.0 \x1b(B\x1b[m\x1b[39;49mst\x1b(B\x1b[m\x1b[39;49m\x1b(B\x1b[m\x1b[39;49m\x1b[K\n%Cpu2  : ....

Reply:
@huangyum You are right, try simple string it works well, thanks.